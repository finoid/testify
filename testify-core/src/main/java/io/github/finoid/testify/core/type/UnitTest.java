package io.github.finoid.testify.core.type;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

/**
 * Base test class for unit-related tests, providing JSON snapshot testing, and component setup.
 */
@Tag("UnitTest")
@ExtendWith({MockitoExtension.class})
public class UnitTest {

}
