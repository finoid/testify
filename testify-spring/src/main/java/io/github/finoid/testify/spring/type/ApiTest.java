package io.github.finoid.testify.spring.type;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.finoid.testify.spring.http.HttpAsserter.HttpAsserterDsl;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJsonTesters;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit.jupiter.SpringExtension;

/**
 * Base test class for API-related tests, providing utility methods for HTTP assertions and component setup.
 */
@Tag("ApiTest")
@AutoConfigureJsonTesters
@SuppressWarnings({"SpringJavaInjectionPointsAutowiringInspection"})
@ExtendWith({SpringExtension.class, SnapshotExtension.class})
@Import(ApiTest.ApiTestConfiguration.class)
public class ApiTest {
    @TestConfiguration
    public static class ApiTestConfiguration {
        @Bean
        public HttpAsserterDsl httpAsserterDsl(final ObjectMapper objectMapper) {
            return HttpAsserterDsl.defaultConfiguration()
                .objectMapper(objectMapper);
        }
    }
}
