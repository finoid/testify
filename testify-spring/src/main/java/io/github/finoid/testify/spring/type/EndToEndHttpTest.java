package io.github.finoid.testify.spring.type;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.finoid.testify.snapshot.SnapshotterExtension;
import io.github.finoid.testify.spring.http.HttpAsserter;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.test.web.servlet.MockMvc;

/**
 * Base test class for E2E-related tests, providing utility methods for HTTP assertions,
 * JSON snapshot testing, and component setup.
 */
@Tag("EndToEndHttpTest")
@SpringBootTest
@AutoConfigureMockMvc
@SuppressWarnings({"SpringJavaInjectionPointsAutowiringInspection"})
@ExtendWith({SnapshotterExtension.class})
@Import(EndToEndHttpTest.EndToEndHttpTestConfiguration.class)
public class EndToEndHttpTest {
    @Autowired
    @SuppressWarnings("UnusedVariable")
    private HttpAsserter httpAsserter;

    @TestConfiguration
    public static class EndToEndHttpTestConfiguration {
        @Bean
        public HttpAsserter httpAsserter(final MockMvc mvc, final ObjectMapper objectMapper) {
            return HttpAsserter.ofMockMvcAndObjectMapper(mvc, objectMapper);
        }
    }
}
