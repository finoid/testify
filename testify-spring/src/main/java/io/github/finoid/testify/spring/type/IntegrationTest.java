package io.github.finoid.testify.spring.type;

import io.github.finoid.testify.snapshot.SnapshotterExtension;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.context.SpringBootTest;

/**
 * Base class for integration tests, testing utilities.
 */
@Tag("IntegrationTest")
@SpringBootTest
@ExtendWith({SnapshotterExtension.class})
public class IntegrationTest {

}
