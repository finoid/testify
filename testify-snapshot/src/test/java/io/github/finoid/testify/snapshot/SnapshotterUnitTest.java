package io.github.finoid.testify.snapshot;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.SnapshotVerifier;
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import io.github.finoid.testify.core.type.UnitTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;

class SnapshotterUnitTest extends UnitTest {
    @Test
    void givenRecordInstanceToBeSerialized_whenJsonSnapshot_thenSuccessfulJsonSnapshotTaken(TestInfo testInfo) {
        var snapshotter = givenSnapshotVerifier(testInfo);

        snapshotter.json()
            .snapshot(new User(1, "John"));
    }

    @Test
    void givenRecordInstanceToBeSerialized_whenPlainSnapshot_thenSuccessfulPlainSnapshotTaken(TestInfo testInfo) {
        var snapshotter = givenSnapshotVerifier(testInfo);

        snapshotter.plain()
            .snapshot(new User(1, "John"));
    }

    @Test
    void givenRecordInstanceToBeSerialized_whenBase64Snapshot_thenSuccessfulBase64SnapshotTaken(TestInfo testInfo) {
        var snapshotter = givenSnapshotVerifier(testInfo);

        snapshotter.base64()
            .snapshot(new User(1, "John"));
    }

    @Test
    void xml() {
        // TODO (nw) to be implemented
    }

    private Snapshotter givenSnapshotVerifier(final TestInfo testInfo) {
        final SnapshotVerifier verifier = new SnapshotVerifier(new PropertyResolvingSnapshotConfig(), getClass(), true);

        final Expect expect = Expect.of(verifier, testInfo.getTestMethod()
            .orElseThrow(() -> new SnapshotMatchException("Unable to locate test method")));

        return new Snapshotter(expect);
    }

    record User(int id, String name) {
    }
}