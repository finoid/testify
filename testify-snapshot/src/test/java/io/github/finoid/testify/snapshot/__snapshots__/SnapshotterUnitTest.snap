io.github.finoid.testify.snapshot.SnapshotterUnitTest.givenRecordInstanceToBeSerialized_whenBase64Snapshot_thenSuccessfulBase64SnapshotTaken=[
VXNlcltpZD0xLCBuYW1lPUpvaG5d
]


io.github.finoid.testify.snapshot.SnapshotterUnitTest.givenRecordInstanceToBeSerialized_whenJsonSnapshot_thenSuccessfulJsonSnapshotTaken=[
  {
    "id" : 1,
    "name" : "John"
  }
]


io.github.finoid.testify.snapshot.SnapshotterUnitTest.givenRecordInstanceToBeSerialized_whenPlainSnapshot_thenSuccessfulPlainSnapshotTaken=[
User[id=1, name=John]
]